trace_decoder/src/processed_block_trace_smt.rs

```diff
@@ -0,0 +1,84 @@
+use std::collections::HashMap;
+
+use ethereum_types::H256;
+use ethereum_types::{Address, U256};
+
+use crate::types::CodeHash;
+use crate::{
+    aliased_crate_types::SmtGenerationInputs,
+    code_hash_resolver::CodeHashResolver,
+    decoding_smt::SmtTraceParsingResult,
+    processed_block_trace::ProcessedBlockTrace,
+    trace_protocol::BlockTrace,
+    types::{CodeHashResolveFunc, OtherBlockData},
+};
+pub(crate) type SmtProcessedBlockTrace = ProcessedBlockTrace<ProcedBlockTraceSmtSpec>;
+
+/// Smt processed pre-image.
+#[derive(Clone, Debug)]
+pub struct SmtProcessedBlockTracePreImage {}
+
+#[derive(Debug)]
+pub(crate) struct ProcedBlockTraceSmtSpec {}
+
+#[allow(dead_code)]
+struct SMTCodeHashResolving<F> {
+    /// If we have not seen this code hash before, use the resolve function that
+    /// the client passes down to us. This will likely be an rpc call/cache
+    /// check.
+    client_code_hash_resolve_f: F,
+
+    /// Code hash mappings that we have constructed from parsing the block
+    /// trace. If there are any txns that create contracts, then they will also
+    /// get added here as we process the deltas.
+    extra_code_hash_mappings: HashMap<CodeHash, Vec<u8>>,
+}
+
+impl<F: CodeHashResolveFunc> CodeHashResolver for SMTCodeHashResolving<F> {
+    fn resolve(&mut self, c_hash: &CodeHash) -> Vec<u8> {
+        match self.extra_code_hash_mappings.get(c_hash) {
+            Some(code) => code.clone(),
+            None => (self.client_code_hash_resolve_f)(c_hash),
+        }
+    }
+
+    fn insert_code(&mut self, c_hash: H256, code: Vec<u8>) {
+        self.extra_code_hash_mappings.insert(c_hash, code);
+    }
+}
```

Comment:
Hmm I don't think that we should be providing implementations of `CodeHashResolver`. What I was thinking is this comes from the client, where `resolve` would hit some internal cache or make some rpc calls if it's missing. `insert_code` is meant to be called when the decoder finds a code mapping within the embedded pre-image.

Reply:
Lets discuss more on this to be a bit more clear on the implementation